### PR TITLE
Added OBJ loader for the ModelLoaderRegistry system.

### PR DESCRIPTION
**Pull request by shadekiller666**
Starting to experiment with using custom data from the forge blockstates json. Everything else is working as intended.

### CODE REVIEW

**Comment by tambry** [comment]
Any news on, why isn't it merged? Still waiting for this to update my mod to 1.8.

**Comment by LexManos** [comment]
@shadekiller666 and @RainWarrior are in charge here. It's their baby.

**Comment by shadekiller666** [comment]
I'm waiting for @RainWarrior to let me know if theres anything major with the OBJLoader that needs fixing, he's ultimately the one who will pull it in. @tambry in the mean time, you can copy OBJLoader and OBJModel into your own mod, and register them with "ModelLoaderRegistry.registerLoader(OBJLoader.instance)", and then use the example mods on this repo to figure out how to implement what you need. The example you want is ModelLoaderRegistryDebug. CustomModelBlock2-whatever the last CustomModelBlock is in that file are the ones that use the OBJ loader in various ways.

**Comment by AchielC** [comment]
@shadekiller666 I copied the OBJLoader and OBJModel to my mod and it's working smoothly. The only problem I encountered is that my texture files need to be mirrored vertically to match my uv-mapping in blender. I don't know if that's intentional.

**Comment by shadekiller666** [comment]
Blender has always been weird about uvs and OBJ files in general, the loader applies uvs properly, your uvs are likely flipped in Blender. Either way that is something that is incorrect with your model, not the loader :P.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java [hunk 1]
```diff
@@ -214,6 +214,17 @@ protected ResourceLocation getModelLocation(ResourceLocation model)
     {
         return new ResourceLocation(model.getResourceDomain(), model.getResourcePath() + ".json");
     }
+    
+    @Override
+    public ModelBlockDefinition getModelBlockDefinition(ResourceLocation location)
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java]: Unused, as far as I can see

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java [hunk 2]
```diff
@@ -214,6 +214,17 @@ protected ResourceLocation getModelLocation(ResourceLocation model)
...
+    {
+        return super.getModelBlockDefinition(location);
+    }
+    
+    public ResourceLocation getBlockStateLocation(ResourceLocation location)
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java]: Same as above

File: src/main/java/net/minecraftforge/client/model/TRSRTransformation.java [hunk 3]
```diff
@@ -68,6 +68,27 @@ public TRSRTransformation(ModelRotation rotation)
...
+    }
+    
+    public static Matrix4f getMatrix(EnumFacing facing)
+    {
+        Matrix4f ret = new Matrix4f();
+        switch(facing)
+        {
+        case DOWN: ret = ModelRotation.X90_Y0.getMatrix(); break;
+        case UP: ret = ModelRotation.X270_Y0.getMatrix(); break;
+        case NORTH: ret = ModelRotation.X0_Y0.getMatrix(); break;
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/TRSRTransformation.java]: TRSRTransformation.identity would probably be better here. And new matrix only needs to be created in the "default" case.

File: src/main/java/net/minecraftforge/client/model/TRSRTransformation.java [hunk 4]
```diff
@@ -516,4 +537,54 @@ public static TRSRTransformation blockCornerToCenter(TRSRTransformation transfor
         ret.mul(tmp);
         return new TRSRTransformation(ret);
     }
+
+    @Override
+    public int hashCode()
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/TRSRTransformation.java]: only matrix should be used for equals/hashcode, all other fields don't count for identity comparisons; don't rely too much on IDE method generators.

File: src/main/java/net/minecraftforge/common/property/Properties.java [hunk 5]
```diff
@@ -6,9 +6,11 @@
 
 import org.apache.commons.lang3.reflect.ConstructorUtils;
 
+import com.google.common.base.Optional;
+
 public class Properties
 {
-    public static <P extends IProperty, V>IUnlistedProperty<V> toUnlisted(P property)
+    public static <P extends IProperty, V>IUnlistedProperty toUnlisted(P property)
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/common/property/Properties.java]: Why is this needed? And all the other places with added/removed casts/generics?

Comment (shadekiller666) [inline src/main/java/net/minecraftforge/common/property/Properties.java]: If I remember correctly, this should be like it was previously... I will double check.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java (hunk 1 above)
Comment (shadekiller666) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java]: This was added to expose ModelBakery.getModelBlockDefinition() as it is protected, to make lives easier should anyone need it.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java (hunk 2 above)
Comment (shadekiller666) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java]: This was added as a means to easily get the resource location that points to the blockstate folder.

File: src/main/java/net/minecraftforge/client/model/TRSRTransformation.java (hunk 4 above)
Comment (shadekiller666) [inline src/main/java/net/minecraftforge/client/model/TRSRTransformation.java]: This has been addressed.

File: .gitignore [hunk 6]
```diff
@@ -15,6 +15,7 @@
 /.gradle
 
 #occupational hazards
+/eclipse/
```
Comment (RainWarrior) [inline .gitignore]: redundant, line 7

File: src/main/java/net/minecraftforge/client/model/ModelFluid.java [hunk 7]
```diff
@@ -132,11 +132,11 @@ public BakedFluid(TRSRTransformation transformation, VertexFormat format, int co
                 {
                     if(gas)
                     {
-                        y[i] = 1 - state.getValue(BlockFluidBase.LEVEL_CORNERS[i]);
+                        y[i] = 1 - (Float)state.getValue(BlockFluidBase.LEVEL_CORNERS[i]);
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/ModelFluid.java]: Why is this needed?

File: src/main/java/net/minecraftforge/client/model/b3d/B3DLoader.java [hunk 8]
```diff
@@ -61,6 +61,7 @@
 import org.lwjgl.BufferUtils;
 
 import com.google.common.base.Function;
+import com.google.common.base.Optional;
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/b3d/B3DLoader.java]: Unnecessary

File: src/main/java/net/minecraftforge/client/model/b3d/B3DLoader.java [hunk 9]
```diff
@@ -299,7 +300,7 @@ public boolean isValid(B3DState value)
             return value instanceof B3DState;
         }
 
-        public Class<B3DState> getType()
+        public Class getType()
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/model/b3d/B3DLoader.java]: Raw type here, generic in the base

File: src/main/java/net/minecraftforge/common/property/Properties.java [hunk 10]
```diff
@@ -6,6 +6,8 @@
 
 import org.apache.commons.lang3.reflect.ConstructorUtils;
 
+import com.google.common.base.Optional;
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/common/property/Properties.java]: Unnecessary

File: src/main/java/net/minecraftforge/common/property/PropertyFloat.java [hunk 11]
```diff
@@ -1,5 +1,6 @@
 package net.minecraftforge.common.property;
 
+import com.google.common.base.Optional;
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/common/property/PropertyFloat.java]: Do you look at the diff when after you commit?

File: src/main/java/net/minecraftforge/common/property/PropertyFloat.java [hunk 12]
```diff
@@ -29,7 +30,7 @@ public boolean isValid(Float value)
         return validator.apply(value);
     }
 
-    public Class<Float> getType()
+    public Class getType()
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/common/property/PropertyFloat.java]: Same as higher up

**Comment by RainWarrior** [comment]
Since github doesn't show the diff for OBJModel.java, I'll provide the comments here:

[getLocation](https://github.com/shadekiller666/MinecraftForge/blob/63d59c24ce0547b73c4e2ba06ee6cb8b59b14029/src/main/java/net/minecraftforge/client/model/obj/OBJModel.java#L144-L149) is unused.

[This](https://github.com/shadekiller666/MinecraftForge/blob/63d59c24ce0547b73c4e2ba06ee6cb8b59b14029/src/main/java/net/minecraftforge/client/model/obj/OBJModel.java#L171) is clearly wrong; remapping the texture in one place will change it in all uses of the model (since the material library is shared).

**Comment by tambry** [comment]
@RainWarrior Any news/progress on getting this merged? Still waiting for this to port my mod to 1.8.

**Comment by shadekiller666** [comment]
@tambry we're working on it, @RainWarrior hasn't gone through the code enough to fully understand how it works and isn't comfortable with pulling it in until he is

**Comment by shadekiller666** [comment]
@tambry about porting your mod, feel free to distribute the OBJLoader and OBJModel classes with your mod until forge has them pulled in

**Comment by shadekiller666** [comment]
To anyone who wishes to use the OBJLoader before its pulled, follow these steps:

1: Make sure you have the latest version of forge.
2: Copy OBJLoader.java and OBJModel.java into their own package in your mod.
3: Copy the `public static Matrix4f getMatrix(EnumFacing facing)` method on line 77 of TRSRTransformation.java, and put it under the `public TRSRTransformation apply(IModelPart part)` method in OBJState (line 1134 of OBJModel.java), then change line 1137 to `return new TRSRTransformation(getMatrix(EnumFacing.NORTH));`
4: During the FMLPreInitializationEvent phase, in the ClientProxy of your mod, register the OBJLoader by calling `ModelLoaderRegistry.registerLoader(OBJLoader.instance);`
5: On the next line in ClientProxy, call `OBJLoader.instance.addDomain(<modid>.toLowerCase());` where <modid> is the modid defined in your main mod class.

**Comment by RainWarrior** [comment]
Merged: https://github.com/MinecraftForge/MinecraftForge/commit/d1b220f6a375c64c0172137b8e08b4827450298d

**Comment by tambry** [comment]
@shadekiller666 Hi.
I'm finally porting my mod to 1.8, but I can't figure out how to properly load .obj model, that would replace a vanilla item model.
Currently my code for loading it is like [this](http://puu.sh/l5xyp/f711adb209.png).
In-game the apple model looks like [this](http://puu.sh/l5vJd/1e22fea2cf.png).

**Comment by Spartan322** [comment]
@tambry That should be a forum post, your not suppose to ask that here.

**Comment by tambry** [comment]
@Spartan322 It seemed like the most reasonable place to reach him. As I'm a bit doubtful, that he visits forums that often.

**Comment by shadekiller666** [comment]
@tambry hmm, not exactly sure how to do this, i know it is possible, just not sure how. 1st thing you should do is: remove the ModelLoaderRegistry.registerLoader line in the code you linked, as the objloader is now in forge (make sure you are using the latest forge). Now, about the actual replacement of the apple model... this may require a custom state mapper. I am currently not at my pc with the dev env, so as soon as i am i will try things out and get back to you. Also, youre right, i dont visit the forums that often, this is a better means of contacting me, you could also contact me in the #minecraftforge irc channel, as i am usually logged into it, my nickname is the same as it is here.

**Comment by tambry** [comment]
@shadekiller666 I messaged you here, because I couldn't see you on the IRC for a couple of hours. Thanks for looking into it. In 1.7.* I used IItemRenderer to achieve the functionality of overriding the rendering of certain items. In 1.8 it's deprecated and not usable. I'm also afraid I wont be able to properly override rendering of certain item states, without it (like the bow or the fishing rod).

**Comment by shadekiller666** [comment]
Whenever im not in irc it means im not on my pc, so send me a message here and ill respond when i can. I dont know what it is you wish to do with overriding the vanilla item models, but i think things like the bow and fishing rod would require a custom state mapper, assuming they arent hardcoded (like maps), i will take a look when i can.
